coil-base/src/main/java/coil/memory/MemoryCache.kt

```diff
@@ -1,251 +1,63 @@
-    }
-
-    override fun size() = 0
-
-    override fun maxSize() = 0
-
-    override fun invalidate(key: Key) {}
-
-    override fun clearMemory() {}
-
-    override fun trimMemory(level: Int) {}
-}
-
-/** A [MemoryCache] implementation backed by an [LruCache]. */
-private class RealMemoryCache(
-    private val weakMemoryCache: WeakMemoryCache,
-    private val referenceCounter: BitmapReferenceCounter,
-    maxSize: Int,
-    private val logger: Logger?
-) : MemoryCache {
-
-    private val cache = object : LruCache<Key, InternalValue>(maxSize) {
-        override fun entryRemoved(
-            evicted: Boolean,
-            key: Key,
-            oldValue: InternalValue,
-            newValue: InternalValue?
-        ) {
-            val isPooled = referenceCounter.decrement(oldValue.bitmap)
-            if (!isPooled) {
-                // Add the bitmap to the WeakMemoryCache if it wasn't just added to the BitmapPool.
-                weakMemoryCache.set(key, oldValue.bitmap, oldValue.isSampled, oldValue.size)
-            }
-        }
-
-        override fun sizeOf(key: Key, value: InternalValue) = value.size
-    }
-
-    override fun get(key: Key) = cache.get(key) ?: weakMemoryCache.get(key)
-
-    override fun set(key: Key, bitmap: Bitmap, isSampled: Boolean) {
-        // If the bitmap is too big for the cache, don't even attempt to store it. Doing so will cause
-        // the cache to be cleared. Instead just evict an existing element with the same key if it exists.
-        val size = bitmap.allocationByteCountCompat
-        if (size > maxSize()) {
-            val previous = cache.remove(key)
-            if (previous == null) {
-                // If previous != null, the value was already added to the weak memory cache in LruCache.entryRemoved.
-                weakMemoryCache.set(key, bitmap, isSampled, size)
-            }
-            return
-        }
-
-        referenceCounter.increment(bitmap)
-        cache.put(key, InternalValue(bitmap, isSampled, size))
-    }
-
-    override fun size() = cache.size()
-
-    override fun maxSize() = cache.maxSize()
-
-    override fun invalidate(key: Key) {
-        logger?.log(TAG, Log.VERBOSE) { "invalidate, key=$key" }
-        cache.remove(key)
-    }
-
-    override fun clearMemory() {
-        logger?.log(TAG, Log.VERBOSE) { "clearMemory" }
-        cache.trimToSize(-1)
-    }
-
-    override fun trimMemory(level: Int) {
-        logger?.log(TAG, Log.VERBOSE) { "trimMemory, level=$level" }
-        if (level >= TRIM_MEMORY_BACKGROUND) {
-            clearMemory()
-        } else if (level in TRIM_MEMORY_RUNNING_LOW until TRIM_MEMORY_UI_HIDDEN) {
-            cache.trimToSize(size() / 2)
-        }
-    }
-
-    private class InternalValue(
-        override val bitmap: Bitmap,
-        override val isSampled: Boolean,
-        val size: Int
-    ) : Value
-
-    companion object {
-        private const val TAG = "RealMemoryCache"
+@ExperimentalCoilApi
+interface MemoryCache {
```

Comment:
The existing `MemoryCache` on master was renamed to `StrongMemoryCache` (because it holds strong references). `MemoryCache` is now a public interface that wraps `StrongMemoryCache` + `WeakMemoryCache`.